### PR TITLE
SLVS-855 S3597 - "ServiceContract" and "OperationContract" attributes…

### PR DESCRIPTION
**Pull request by tvsonar**
… should be used together

### CODE REVIEW

File: src/SonarLint.CSharp/Rules/ArgumentSpecifiedForCallerInfoParameter.cs [hunk 1]
```diff
@@ -100,14 +101,13 @@ protected override void Initialize(SonarAnalysisContext context)
         private static AttributeData GetCallerInfoAttribute(IParameterSymbol parameter)
         {
             return parameter.GetAttributes()
-                .FirstOrDefault(attr => CallerInfoAttributeNames.Contains(attr.AttributeClass.ToDisplayString()));
+                .FirstOrDefault(attr => attr.AttributeClass.IsAny(CallerInfoAttributeTypes));
```
Comment (tvsonar) [inline src/SonarLint.CSharp/Rules/ArgumentSpecifiedForCallerInfoParameter.cs]: This change should be in a separate PR or at least a separate commit.

File: src/SonarLint.CSharp/Rules/WcfMissingContractAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+using SonarLint.Common.Sqale;
+using SonarLint.Helpers;
+
+namespace SonarLint.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [SqaleConstantRemediation("2min")]
+    [SqaleSubCharacteristic(SqaleSubCharacteristic.UsabilityAccessibility)]
+    [Rule(DiagnosticId, RuleSeverity, Title, IsActivatedByDefault)]
+    [Tags(Tag.Convention, Tag.Suspicious)]
```
Comment (tvsonar) [inline src/SonarLint.CSharp/Rules/WcfMissingContractAttribute.cs]: These have been changed in the spec

File: src/SonarLint.CSharp/Rules/WcfMissingContractAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+                    }
+
+                    var classOrInterface = GetMessagePartClassOrInterface(namedType);
+
+                    string messageEnding;
+                    string attributeToAdd;
+
+                    if (hasServiceContract && !hasAnyMethodWithOperationContract)
+                    {
+                        messageEnding = string.Format(MessageOperation, classOrInterface);
```
Comment (tvsonar) [inline src/SonarLint.CSharp/Rules/WcfMissingContractAttribute.cs]: it would be enough to select the message ending here, and do the `string.format` after the `if-else`.

File: src/SonarLint.CSharp/Rules/WcfMissingContractAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+        private static TypeDeclarationSyntax GetTypeDeclaration(INamedTypeSymbol namedType, Compilation compilation)
+        {
+            return namedType.DeclaringSyntaxReferences
+                .Where(sr => !sr.SyntaxTree.IsGenerated(compilation))
+                .Select(sr => sr.GetSyntax() as TypeDeclarationSyntax)
+                .FirstOrDefault(s => s != null);
+        }
+
+        private static string GetMessagePartClassOrInterface(INamedTypeSymbol namedType)
```
Comment (tvsonar) [inline src/SonarLint.CSharp/Rules/WcfMissingContractAttribute.cs]: This method is only used in one place, the code would be more readable with inlining the method body there.

File: src/SonarLint.CSharp/Rules/WcfMissingContractAttribute.cs [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+                    {
+                        return;
+                    }
+
+                    var classOrInterface = GetMessagePartClassOrInterface(namedType);
+
+                    string messageEnding;
+                    string attributeToAdd;
+
+                    if (hasServiceContract && !hasAnyMethodWithOperationContract)
```
Comment (dbolkensteyn) [inline src/SonarLint.CSharp/Rules/WcfMissingContractAttribute.cs]: `!hasAnyMethodWithOperationContract` seems redundant here given the xor on line 76

**Comment by dbolkensteyn** [comment]
LGTM
